cores/nRF5/WInterrupts.c

```diff
@@ -148,30 +192,39 @@ void GPIOTE_IRQHandler()
   SEGGER_SYSVIEW_RecordEnterISR();
 #endif
 
-  uint32_t event = offsetof(NRF_GPIOTE_Type, EVENTS_IN[0]);
-
+  // Read this once (not 8x), as it's a volatile read
+  // across the AHB, which adds up to 3 cycles.
+  uint32_t const enabledInterruptMask = NRF_GPIOTE->INTENSET;
   for (int ch = 0; ch < NUMBER_OF_GPIO_TE; ch++) {
-    if ((*(uint32_t *)((uint32_t)NRF_GPIOTE + event) == 0x1UL) && (NRF_GPIOTE->INTENSET & (1 << ch))) {
-      if (channelMap[ch] != -1 && callbacksInt[ch]) {
-        if ( callbackDeferred[ch] )  {
-          // Adafruit defer callback to non-isr if configured so
-          ada_callback(NULL, 0, callbacksInt[ch]);
-        }else{
-         callbacksInt[ch]();
-        }
+    // only process where the interrupt is enabled and the event register is set
+    // check interrupt enabled mask first, as already read that IOM value, to
+    // reduce delays from AHB (16MHz) reads.
+    if ( 0 == (enabledInterruptMask & (1 << ch))) continue;
+    if ( 0 == NRF_GPIOTE->EVENTS_IN[ch]) continue;
+
+    // If the event was set and interrupts are enabled,
+    // call the callback function only if it exists,
+    // but ALWAYS clear the event to prevent an interrupt storm.
+    if (channelMap[ch] != -1 && callbacksInt[ch]) {
+      if ( callbackDeferred[ch] ) {
+        // Adafruit defer callback to non-isr if configured so
+        ada_callback(NULL, 0, callbacksInt[ch]);
+      } else {
+        callbacksInt[ch]();
       }
-
-    *(uint32_t *)((uint32_t)NRF_GPIOTE + event) = 0;
-#if __CORTEX_M == 0x04
-    volatile uint32_t dummy = *((volatile uint32_t *)((uint32_t)NRF_GPIOTE + event));
-    (void)dummy;
-#endif
     }
 
-    event = (uint32_t)((uint32_t)event + 4);
+    // clear the event
+    NRF_GPIOTE->EVENTS_IN[ch] = 0;
   }
+#if __CORTEX_M == 0x04
+  // See note at nRF52840_PS_v1.1.pdf section 6.1.8 ("interrupt clearing")
+  // See also https://gcc.gnu.org/onlinedocs/gcc/Volatiles.html for why
+  // using memory barrier instead of read of an unrelated volatile
+  __DSB(); __NOP();__NOP();__NOP();__NOP();
```

Comment:
thanks, this is much better than the original read then discard register. I think this is the only place need adding the _DSB()